### PR TITLE
feat(inverted_index.search): add index applier

### PR DESCRIPTION
**Pull request by zhongzc**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Introduce `IndexApplier` to apply the predefined predicates to the data read by the given index reader, returning a list of relevant indices.

Applier instances are reusable and work with various `InvertedIndexReader` instances, avoiding repeated compilation of fixed predicates such as regex patterns.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
Prerequisite merging https://github.com/GreptimeTeam/greptime-proto/pull/126

https://github.com/GreptimeTeam/greptimedb/issues/2705

### CODE REVIEW

File: src/index/src/inverted_index/search/index_apply/predicates_apply.rs [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+        let metadata = reader.metadata().await?;
+
+        let mut bitmap = Self::bitmap_full_range(&metadata);
+        // TODO(zhongzc): optimize the order of applying
+        for (tag_name, fst_applier) in &self.fst_appliers {
+            if bitmap.count_ones() == 0 {
+                break;
+            }
+
+            // TODO(zhongzc): how to handle missing columns?
```
[the remark this answers is not in the document]
Comment (zhongzc) [reply src/index/src/inverted_index/search/index_apply/predicates_apply.rs:50]: Good suggestion 👍

File: src/index/src/inverted_index/search/index_apply/predicates_apply.rs [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Default)]
+pub struct ReadOptions {
+    /// `index_not_found_strategy` controls the behavior of the applier when the index is not found.
+    pub index_not_found_strategy: IndexNotFoundStrategy,
+}
+
+/// Defines the behavior of an applier when the index is not found.
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Default)]
+pub enum IndexNotFoundStrategy {
+    /// If the index is not found, the applier will return an empty list of indices.
```
Comment (killme2008) [inline src/index/src/inverted_index/search/index_apply/predicates_apply.rs:135]: Could simplify the description of these variants. The struct already describes its purpose. For example:

```rust
/// Return an empty list of indices.
ReturnEmpty,

/// Return a list of all indices.
ReturnFullRange
...
}
```

**Review by killme2008 — COMMENTED** [review]
Good job! A few suggestions attached.
Another suggestion: maybe we can add some benchmarks for inverted index performance to get a baseline and prepare for future improvements.

**Comment by zhongzc** [comment]
> Another suggestion: maybe we can add some benchmarks for inverted index performance to get a baseline and prepare for future improvements.

Agreed. I will construct some bench cases with local files as the baseline in the upcoming PRs.

**Review by killme2008 — APPROVED** [review]
LGTM

File: src/index/src/inverted_index/search/index_apply/predicates_apply.rs [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+
+        let mut bitmap = Self::bitmap_full_range(&metadata);
+        // TODO(zhongzc): optimize the order of applying to make it quicker to return empty.
+        for (name, fst_applier) in &self.fst_appliers {
+            if bitmap.count_ones() == 0 {
+                break;
+            }
+
+            let Some(meta) = metadata.metas.get(name) else {
+                match self.options.index_not_found_strategy {
```
Comment (evenyag) [inline src/index/src/inverted_index/search/index_apply/predicates_apply.rs:61]: Can we provide the options via the `SearchContext`?

Comment (zhongzc) [reply src/index/src/inverted_index/search/index_apply/predicates_apply.rs:61]: This is a very marginal decision. I think it depends on whether there will be an applier implementation that doesn't care about this configuration item in the future. If present, this change would be disruptive. But you know, I'm struggling. But now that you have this comment, I probably don't need to worry about it anymore, let's do it

File: src/index/src/inverted_index/search/index_apply/predicates_apply.rs [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+                    IndexNotFoundStrategy::ReturnEmpty => {
+                        return Ok(vec![]);
+                    }
+                    IndexNotFoundStrategy::ReturnFullRange => {
```
Comment (evenyag) [inline src/index/src/inverted_index/search/index_apply/predicates_apply.rs:65]: What about `Ignore` or `Skip`? We want the applier to ignore this column. It doesn't return.
```suggestion
                    IndexNotFoundStrategy::Ignore => {
```

**Review by evenyag — APPROVED** [review]
Rest LGTM

**Pull request merged**